### PR TITLE
Instance port allocation for CloudFormation backend

### PR DESCRIPTION
**Pull request by ejholmes**
Addresses https://github.com/remind101/empire/pull/803#discussion_r61543193

This adds instance port allocation to the CloudFormation scheduling backend by using CloudFormation custom resources.

In a nutshell, when we provision a CloudFormation stack for an app, every container port in a task definition gets a `Custom::InstancePort` resource (e.g. `web8080InstancePort`). CloudFormation sends the provisioning request to an SNS topic, and Empire pulls these off of an SQS queue and provisions the ports (or releases them) using the existing port allocator.

This also sets up a nice little framework for adding [additional custom resources](https://github.com/remind101/empire/issues/810).

**TODO**

* [ ] Auto-subscribe sqs queue to sns queue
* [ ] Don't start custom resources server if sqs queue isn't set.
* [ ] Error out of cloudformation scheduler is sqs queue isn't set.

### CODE REVIEW

**Comment by ejholmes** [comment]
Also, note that this is getting merged into the `cloudformation` branch, not master.

**Comment by ejholmes** [comment]
On a side note, this might make for a good blog post, since this is a pretty nice general way of dynamically handling the instance port limitations of ECS.

File: cmd/empire/server.go [hunk 1]
```diff
@@ -25,11 +26,26 @@ func runServer(c *cli.Context) {
...
+	e, err := newEmpire(db, c)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	r, err := newReporter(c)
+	if err != nil {
+		log.Fatal(err)
+	}
+	p := cloudformation.NewCustomResourceProvisioner(db.DB.DB(), newConfigProvider(c))
```
Comment (ejholmes) [inline cmd/empire/server.go]: This should only actually start if `FlagCustomResourcesQueue` is set.

**Comment by mhahn** [comment]
👍 this is really cool.

This relies on an SNS topic and an SQS topic that is subscribed to the SNS topic correct?

**Comment by ejholmes** [comment]
@mhahn yep. I should update the example CloudFormation stack to create the subscription automatically.

**Comment by ejholmes** [comment]
I'll merge this into the cloudformation branch and continue with any changes there.

**Pull request merged**